refactor(forms): add onError callback to validateHttp for HTTP errors

napulitanfrontend opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
is not possible handle server errors when validateHttp is applied on a field in a signal form based:

```typescript
f = form(signal({ name: '' }), (p) => {
    required(p.name, { message: 'name is required!' }),
      validateHttp(p.name, {
        request: ({ value }) => {
          return value() ? `https://api.github.com/users/${value()}` : undefined;
        },
        errors: (res: any) => {
          return res && res.id && !res.message
            ? [customError({ kind: 'notUnique', message: "l'utente esiste già" })]
            : [];
        }
      });
  });
  ```
  if server responde with http status different from 200 errors callback will not never called
  
## What is the new behavior?

now, throught onError configuration is possible to handle also server errors and different http status

```typescript
f = form(signal({ name: '' }), (p) => {
    required(p.name, { message: 'name is required!' }),
      validateHttp(p.name, {
        request: ({ value }) => {
          return value() ? `https://api.github.com/users/${value()}` : undefined;
        },
        errors: (res: any) => {
          return res && res.id && !res.message
            ? [customError({ kind: 'notUnique', message: "l'utente esiste già" })]
            : [];
        },
        onError(error:any, ctx) {
              return error && error.message
            ? [customError({ kind: 'serverError', message: error.message })]
            : [];
        },
      });
  });

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

thePunderWoman commented:
@napulitanfrontend Thanks for the contribution. You'll have to sign the CLA before we can move forward with this.

napulitanfrontend commented:
> @napulitanfrontend Thanks for the contribution. You'll have to sign the CLA before we can move forward with this.

@thePunderWoman 
it seems i'v already sign the CLA

napulitanfrontend commented:
> @napulitanfrontend It appears you have, but your email does not match what you signed with. So the CLA rule is failing. You'd need to also sign it with your email. See https://github.com/angular/angular/pull/63949/checks?check_run_id=50806823787 to see the failure.

Got It, signed CLA with different email address

kirjs commented on packages/forms/signals/test/node/validation_status.spec.ts:
```diff
@@ -383,7 +388,8 @@ describe('validation status', () => {
                   ),
               })),
-            errors: (errs) => errs,
+            onSuccess: (errs) => errs as TreeValidationResult,
```
nit: just return empty array here and below?

kirjs commented:
please resolve all the errors

kirjs commented:
Looks like this is still broken, can you drop the lock file?

JeanMeche commented:
Drop both lock files*

kirjs reviewed (approved):
This is approved, please fix the lint warnings.

mmalerba commented on packages/forms/signals/test/node/validation_status.spec.ts:
```diff
@@ -188,7 +188,8 @@ describe('validation status', () => {
                   ),
               })),
-            errors: (errs) => errs,
+            onSuccess: (errs) => errs,
```
nit: can you name the parameter `result` (I understand it is an array of `ValidationError` in this case, but its confusing to see the `onSuccess` taking `errs` as a parameter). Same comment for the other tests.

mmalerba commented:
Please run `pnpm bazel run //packages/forms:forms_api.accept` to update the API goldens

mmalerba reviewed (approved):
reviewed-for: public-api

Commit:
refactor(forms): add onError callback to validateHttp for HTTP errors

Adds onError callback inside validateHttp validator in signal forms

PR-Close: #63949

JeanMeche commented:
Just ran the final set of tests and added the PR to the merge queue ! Thx for this.
